lxd/patches.go

```diff
@@ -952,40 +1007,249 @@ func upgradeFromStorageTypeZfs(name string, d *Daemon, defaultPoolName string, d
+			if err != nil {
+				// In case we're on a LXD instance where a a partial
+				// upgrade took place, we should still proceed.
+				shared.LogWarnf("Could not insert a storage volume for image \"%s\".", img)
+				return err
+			}
+		} else {
+			shared.LogErrorf("Failed to query database: %s", err)
+			return err
 		}
 
 		imageMntPoint := getImageMountPoint(poolName, img)
-		err = os.MkdirAll(imageMntPoint, 0700)
-		if err != nil {
-			return err
+		if !shared.PathExists(imageMntPoint) {
+			err := os.MkdirAll(imageMntPoint, 0700)
+			if err != nil {
+				shared.LogWarnf("Failed to create image mountpoint. Proceeding...")
+			}
 		}
 
 		oldImageMntPoint := shared.VarPath("images", img+".zfs")
 		imageDataset := fmt.Sprintf("%s/images/%s", defaultPoolName, img)
-		if shared.PathExists(oldImageMntPoint) {
-			if shared.IsMountPoint(oldImageMntPoint) {
-				output, err := tryExec("zfs", "unmount", "-f", imageDataset)
+		if shared.PathExists(oldImageMntPoint) && shared.IsMountPoint(oldImageMntPoint) {
+			_, err := tryExec("zfs", "unmount", "-f", imageDataset)
+			if err != nil {
+				shared.LogWarnf("Failed to unmount ZFS filesystem via zfs unmount. Trying lazy umount (MNT_DETACH)...")
+				err := tryUnmount(oldImageMntPoint, syscall.MNT_DETACH)
 				if err != nil {
-					return fmt.Errorf("Failed to unmount ZFS filesystem: %s", output)
+					shared.LogWarnf("Failed to unmount ZFS filesystem: %s", err)
 				}
 			}
 
-			err = os.Remove(oldImageMntPoint)
-			if err != nil {
-				return err
-			}
+			os.Remove(oldImageMntPoint)
 		}
 
 		// Set new mountpoint for the container's dataset it will be
 		// automatically mounted.
 		output, err := exec.Command("zfs", "set", "mountpoint=none", imageDataset).CombinedOutput()
 		if err != nil {
-			return fmt.Errorf("Failed to set new ZFS mountpoint: %s.", output)
+			shared.LogWarnf("Failed to set new ZFS mountpoint: %s.", output)
+		}
+	}
+
+	var finalErr error
+	if len(failedUpgradeEntities) > 0 {
+		finalErr = fmt.Errorf(strings.Join(failedUpgradeEntities, "\n"))
+	}
+
+	return finalErr
+}
+
+func updatePoolPropertyForAllObjects(d *Daemon, poolName string, allcontainers []string) error {
+	// The new storage api enforces that the default storage pool on which
+	// containers are created is set in the default profile. If it isn't
+	// set, then LXD will refuse to create a container until either an
+	// appropriate device including a pool is added to the default profile
+	// or the user explicitly passes the pool the container's storage volume
+	// is supposed to be created on.
+	profiles, err := dbProfiles(d.db)
+	if err == nil {
+		for _, pName := range profiles {
+			pID, p, err := dbProfileGet(d.db, pName)
+			if err != nil {
+				shared.LogErrorf("Could not query database: %s.", err)
+				return err
+			}
+
+			// Check for a root disk device entry
+			k, _ := containerGetRootDiskDevice(p.Devices)
+			if k != "" {
+				// If found, set the pool property
+				p.Devices[k]["pool"] = poolName
```

Comment:
Only do that if it doesn't already have a value set. Since we will trigger this code with 2.9.2, there may be 2.9.1 users that created and configured alternate pools already.